pkg/networking/backend_sg_provider.go

```diff
@@ -76,37 +89,106 @@ type defaultBackendSGProvider struct {
 	ec2Client       services.EC2
 	k8sClient       client.Client
 	logger          logr.Logger
+	objectsMap      sync.Map
+
+	checkIngressFinalizersFunc func([]string) bool
 
 	defaultDeletionPollInterval time.Duration
 	defaultDeletionTimeout      time.Duration
 }
 
-func (p *defaultBackendSGProvider) Get(ctx context.Context) (string, error) {
+func (p *defaultBackendSGProvider) Get(ctx context.Context, activeResources []metav1.Object) (string, error) {
 	if len(p.backendSG) > 0 {
 		return p.backendSG, nil
 	}
 	// Auto generate Backend Security group, and return the id
-	if err := p.allocateBackendSG(ctx); err != nil {
+	if err := p.allocateBackendSG(ctx, activeResources); err != nil {
 		p.logger.Error(err, "Failed to auto-create backend SG")
 		return "", err
 	}
 	return p.autoGeneratedSG, nil
 }
 
-func (p *defaultBackendSGProvider) Release(ctx context.Context) error {
+func (p *defaultBackendSGProvider) Release(ctx context.Context, activeResources []metav1.Object,
+	inactiveResources []metav1.Object, backendSGRequired bool) error {
 	if len(p.backendSG) > 0 {
 		return nil
 	}
+	defer func() {
+		for _, res := range inactiveResources {
+			p.objectsMap.Delete(getObjectKey(res))
+		}
+	}()
+	p.updateObjectsMap(ctx, activeResources, inactiveResources, backendSGRequired)
+	p.logger.V(1).Info("release backend SG", "active", activeResources,
+		"inactive", inactiveResources, "needed", backendSGRequired)
+	if len(activeResources) > 0 && backendSGRequired {
+		return nil
+	}
 	if required, err := p.isBackendSGRequired(ctx); required || err != nil {
 		return err
 	}
 	return p.releaseSG(ctx)
 }
 
-func (p *defaultBackendSGProvider) allocateBackendSG(ctx context.Context) error {
+func (p *defaultBackendSGProvider) updateObjectsMap(_ context.Context, activeResources []metav1.Object,
+	inactiveResources []metav1.Object, backendSGRequired bool) {
+	for _, res := range inactiveResources {
+		p.objectsMap.Store(getObjectKey(res), false)
+	}
+	for _, res := range activeResources {
+		p.objectsMap.Store(getObjectKey(res), backendSGRequired)
+	}
+}
+
+func (p *defaultBackendSGProvider) isBackendSGRequired(ctx context.Context) (bool, error) {
+	var requiredForAny bool
+	p.objectsMap.Range(func(_, v interface{}) bool {
+		if v.(bool) {
+			requiredForAny = true
+			return false
+		}
+		return true
+	})
+	if requiredForAny {
+		return true, nil
+	}
+	if required, err := p.checkIngressListForUnmapped(ctx); required || err != nil {
+		return required, err
```

Comment:
When `required` is true, is it safe to return an `err` of nil? It seems to me in that case it will complete deployment of the model and remove the finalizers from the inactive members. The controller might then stop trying to reconcile the group and leave the SG orphaned.

Reply:
Controller will not remove finalizers in case of error. 

If required is true, then controller skips backend SG cleanup.